### PR TITLE
[MM-28885] app/plugin: check if plugin manifest has errors

### PR DESCRIPTION
**Pull request by isacikgoz**
#### Summary
In such cases there can be manifest errors and it can lead to have a `nil` manifest in the available plugins. We need to check if the the manifest exists.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-28885

### CODE REVIEW

File: app/plugin.go [hunk 1]
```diff
@@ -233,6 +233,9 @@ func (a *App) SyncPlugins() *model.AppError {
 
 	var wg sync.WaitGroup
 	for _, plugin := range availablePlugins {
+		if plugin.Manifest == nil {
```
Comment (lieut-data) [inline app/plugin.go:236]: Should we push this back into `pluginsEnvironment.Available()`, and filter out anything without a manifest? I'd argue "it's not a plugin" without a manifest, so it should probably be suppressed everywhere.

**Review by lieut-data — CHANGES_REQUESTED** [review]
A thought below, but also: can we find a way to unit-test this?

File: app/plugin.go (hunk 1 above)
Comment (isacikgoz) [reply app/plugin.go]: Sure, makes sense 👍

File: plugin/environment_test.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	require.NoError(t, err1)
+	t.Cleanup(func() {
+		os.RemoveAll(dir)
+	})
+
+	env := Environment{
+		pluginDir: dir,
+	}
+
+	t.Run("Should be able to load avaliable plugins", func(t *testing.T) {
```
Comment (lieut-data) [inline plugin/environment_test.go:27]:
```suggestion
 	t.Run("Should be able to load available plugins", func(t *testing.T) { 
```

File: plugin/environment.go [hunk 3]
```diff
@@ -86,9 +86,11 @@ func scanSearchPath(path string) ([]*model.BundleInfo, error) {
 		if !file.IsDir() || file.Name()[0] == '.' {
 			continue
 		}
-		if info := model.BundleInfoForPath(filepath.Join(path, file.Name())); info.ManifestPath != "" {
+		info := model.BundleInfoForPath(filepath.Join(path, file.Name()))
+		if info.Manifest != nil && info.Path != "" {
```
Comment (lieut-data) [inline plugin/environment.go:90]: It's not obvious why we test `info.Path` here -- `BundleInfoForPath` constructs it from the `path` we provide, which necessarily can't be empty here. Should we just test `info.Manifest != nil`?

File: plugin/environment_test.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+		_, err = f.WriteString(bundle1.Manifest.ToJson())
+		require.NoError(t, err)
+		f.Close()
+
+		bundles, err := env.Available()
+		require.NoError(t, err)
+		require.Len(t, bundles, 1)
+	})
+
+	t.Run("Should not be able to load plugins without a valid manifest", func(t *testing.T) {
```
Comment (lieut-data) [inline plugin/environment_test.go:52]: For completeness, can we also test that there's no `manifest.json` at all?

**Review by lieut-data — APPROVED** [review]
Thanks, @isacikgoz! A few additional comments, but no need to block on my review.

File: plugin/environment.go (hunk 3 above)
Comment (isacikgoz) [reply plugin/environment.go]: Makes sense, the check was there already and TBH I don't have much context here so I've left it :)

File: plugin/environment_test.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+		require.NoError(t, err)
+		defer os.RemoveAll(filepath.Join(dir, "plugin1"))
+
+		path := filepath.Join(dir, "plugin1", "plugin.json")
+		f, err := os.Create(path)
+		require.NoError(t, err)
+
+		_, err = f.WriteString(bundle1.Manifest.ToJson())
+		require.NoError(t, err)
+		f.Close()
```
Comment (mgdelacroix) [inline plugin/environment_test.go:45]: nit: only a suggestion, but the process of writing the file can be simplified with [ioutil's WriteFile](https://golang.org/pkg/io/ioutil/#WriteFile)

**Review by mgdelacroix — APPROVED** [review]
LGTM! :+1: 

Added a couple of nonblocking nits

File: plugin/environment_test.go (hunk 5 above)
Comment (isacikgoz) [reply plugin/environment_test.go]: Thanks, nice catch! :)

**Review by lindalumitchell — APPROVED** [review]

**Pull request merged**